### DISCUSSION TITLE
Surfacing arch-specific differences

### DISCUSSION THREAD

**Discussion by sotteson1** [microsoft/win32metadata#418] [Ideas]
**Below is how I'm thinking about solving arch-specific differences. Thoughts, comments?**

To solve #384, we talked about using attributes to surface arch-specific differences. I built a 32-bit version of the .winmd and compared it with the normal 64-bit version using WinmdUtils:

````dotnet D:\repos\win32metadata\sources\WinmdUtils\bin\Release\netcoreapp3.1\WinmdUtils.dll compare --first D:\repos\win32metadata\bin\Windows.Win32.64.winmd --second D:\repos\win32metadata\bin\Windows.Win32.32.winmd````

Some common cases and how I plan to solve them:

* A method or struct in one but not the other: (This one means it's in x64 but not x86)
````Windows.Win32.Debug.Apis.RtlAddFunctionTable(FunctionTable,EntryCount,BaseAddress) not found in 2nd winmd````

    Add this to the metadata:
    ````[ArchDifferences(Supported = Arch.x64)]````

* A struct has a different packing on 32-bit: (This one means 64-bit has no packing but 32-bit does)
````Windows.Win32.WindowsAndMessaging.OPENFILENAMEA :  => [StructLayout(0,Pack=1)]````

    Add this to the metadata:
    ````[ArchDifferences(32BitPacking = 1)]````

* A parameter or return type changes on a method: (This one means the parameter debugApplication changes types between 64 and 32-bit)
````Windows.Win32.Js.Apis.JsCreateContext(runtime,debugApplication,newContext) : debugApplication...IDebugApplication64 => IDebugApplication32````

    This one is trickier, but I think the best solution is to emit it twice, once for each signature, and add attributes to indicate which 
    one is for which:

    ````C#
    [ArchDifferences(Supported = Arch.x64)]
    [DllImport("chakra", ExactSpelling = true)]
    public unsafe static extern JsErrorCode JsCreateContext([In] void* runtime, [In] IDebugApplication32 debugApplication, [Out] void** newContext);

    [ArchDifferences(Supported = Arch.x86)]
    [DllImport("chakra", ExactSpelling = true)]
    public unsafe static extern JsErrorCode JsCreateContext([In] void* runtime, [In] IDebugApplication64 debugApplication, [Out] void** newContext);
  ````

  I'm thinking I would do the same thing for a struct where a field was in one or not the other. I could attribute the field, but I could also attribute the entire struct and emit it twice.

* A really annoying one is where lots of methods and structs in Search use a typedef for an int to be either 32 or 64 bits long depending on the arch: (This means it's a 64-bit int vs a 32-bit int)
````Windows.Win32.Search.DBVECTOR.size...UInt64 => UInt32````

    Original definition:
  
  ````C
  #ifdef _WIN64
  typedef ULONGLONG DBLENGTH;
  #else //_WIN64
  typedef ULONG DBLENGTH;
  #endif

  typedef struct tagDBVECTOR {
	  DBLENGTH size;
	  [size_is((ULONG)size)] void * ptr;
  } DBVECTOR;
  ````

  There are a lot of these. The easiest would be to make these UIntPtr or use a NativeTypedef to make a DBLENGTH that wraps a UIntPtr. That would collapse all of these into the same version so I wouldn't need to add any arch-specific attributes. I wish there was a SIZE_T type in ECMA-335.

**Comment by AArnott** [comment]
What about ARM64? How does that compare?

> A parameter or return type changes on a method

The `JsCreateContext` is an interesting case for two reasons: 
1. I [can't find any documentation for it](https://www.bing.com/search?q=docs.microsoft.com+JsCreateContext&cvid=5240e675bbc14be791da3779ba3b158b&aqs=edge.0.69i59.1779j0j4&FORM=ANAB01&PC=U531).
2. The 64-bit version's type is a COM interface that *may* derive from the COM interface used in the 32-bit version. If that is true, then we can say that the 64-bit version of the function is *always* defined since when `IDebugApplication64` is passed to the 32-bit function, it'll accept the 64-bit interface since it's a derivative, I think.

Anyway I guess my biggest question on this is how common it is to have types changed on methods across architectures.

> I wish there was a SIZE_T type in ECMA-335.

There is: it's `IntPtr`. And in C# 9, use of `nint` or `nuint` actually uses `IntPtr` underneath (in metadata `native int`) but attributes its use in APIs so consumers know to prefer `nint` syntax over `IntPtr`. 
So yes, _please_ find a way to map all these "arch-specific" types that are simply pointer-length integers to `IntPtr`. Then keeping #21 in mind, also attribute that use of `IntPtr` (or use a typedef called `SIZE_T`) to suggest to projections that it should really be a native sized integer, however that is expressed most naturally in the target language.

To *not* coalesce things like `DBLENGTH` into one API definition and rather define an API twice with attributes would lead CsWin32 to refuse to emit these APIs until the project targets specific CPU architectures rather than AnyCPU, which would be a real shame since any hand authored p/invoke signature/type would simply use `IntPtr`/`nint` in order to retain AnyCPU compatibility.

**Reply by kennykerr** [reply]
Both `IDebugApplication64` and `IDebugApplication32` derive from `IRemoteDebugApplication`.

**Comment by kennykerr** [comment]
Rust already maps `IntPtr`/`ELEMENT_TYPE_I` and `UIntPtr`/`ELEMENT_TYPE_U` to Rust's `usize` and `isize` types. No need to mint new pointer-sized integer types. You can find these documented in ECMA-335 under the `ELEMENT_TYPE` constants. 

The `ArchDifferences` attribute seems like a fine solution. I would just keep it simple if possible. So have arch-specific definitions with the `ArchDifferences` attribute telling us which arch each definition applies to, e.g. `[ArchDifferences(x64)`, but avoid more granular information like packing inside this attribute. There is already metadata for that. Simply encode that in the usual way. Then I would just call the attribute `Architecture` as it simply identifies a specific architecture that the definition applies to. I think that would just make it a lot simpler to parse.

**Reply by AArnott** [reply]
> No need to mint new pointer-sized integer types. You can find these documented in ECMA-335 under the ELEMENT_TYPE constants.

What we want to preserve is a semantic distinction between pointer-sized opaque values vs pointer-sized integers, since in some languages like C#, it makes a difference.

**Reply by kennykerr** [reply]
I think by `pointer-sized opaque values` you mean things like handles? I thought that's what the `NativeTypedef` attribute was for. Then when we see a `IntPtr`/`UIntPtr` by itself we know its just a simple pointer-sized integer.

**Reply by AArnott** [reply]
> avoid more granular information like packing inside this attribute

I don't think we want to go into defining the same struct twice in the metadata as that complicates type equivalence wherever that struct is used. Or we could have the attribute include the *real* name of the struct and have all references to that struct refer to the real name. But that would be cumbersome because now you have broken type refs in the metadata.

**Reply by kennykerr** [reply]
Yeah, that's not going to be easy but it sounded like Steve was already suggesting doing that for some of his examples. Maybe I misunderstood.

**Reply by kennykerr** [reply]
Perhaps we can solve this problem with meta structs where there is a single namespace type that contains all the information. 

```C#
[ArchSpecific]
struct Thing
{
    struct x64
    {
        // arch specific fields
    }
    struct x86
    {
        // arch specific fields
    }
}

**Reply by sotteson1** [reply]
oledb.h defines a bunch of integer typedefs that are essentially IntPtr/UIntPtr, like DBLENGTH, DBROWOFFSET, DBOFFSET, and DBCOUNTITEM. I could either turn them all into IntPtr/UIntPtr (@kennykerr , I think that's what you're suggesting), or preserve the typedef and add an attribute to NativeTypedef to show it's an int and not an opaque handle. One thing I like about the typedefs is that they give a little more meaning than just a SIZE_T equivalent, or IntPtr/UIntPtr.

````C#
[NativeTypedef(TreatAsInteger = true)]
public struct DBROWOFFSET
{
    public IntPtr Value;
}
````

**Reply by kennykerr** [reply]
I'm just not sure what the intent would be. Either we "treat as integer" in which case you can just turn them all into `IntPtr`/`UIntPtr`, or we "treat as handle" and you can just use `[NativeTypedef]` with no qualification. 

A Rust developer would write `DBROWOFFSET(123)` for a `NativeTypedef` or simply `123` if its just an integer. It just seems a little gratutious to add types for things like "length" and "offset" when languages like Rust and (modern) C++ just use integers for such things.

**Reply by sotteson1** [reply]
> Perhaps we can solve this problem with meta structs where there is a single namespace type that contains all the information.
> 
> ```cs
> [ArchSpecific]
> struct Thing
> {
>     struct x64
>     {
>         // arch specific fields
>     }
>     struct x86
>     {
>         // arch specific fields
>     }
> }
> ```

So far I haven't implemented anything, so all of this is still up for discussion.

I don't like the idea of adding artificial nested structs. I would rather attribute individual fields to say if they're only in one arch or the other if we're not going to duplicate struct names. @AArnott, I agree that duplicating struct names would mess up references. A struct that has an arch-specific version could include an attribute that points at those arch-specific versions. Duplicating methods shouldn't have the same problem since there won't be references to them.

**Reply by sotteson1** [reply]
> I'm just not sure what the intent would be. Either we "treat as integer" in which case you can just turn them all into `IntPtr`/`UIntPtr`, or we "treat as handle" and you can just use `[NativeTypedef]` with no qualification.
> 
> A Rust developer would write `DBROWOFFSET(123)` for a `NativeTypedef` or simply `123` if its just an integer. It just seems a little gratutious to add types for things like "length" and "offset" when languages like Rust and (modern) C++ just use integers for such things.

I'm fine either way. I don't see anywhere else in the Win32 API that went to the trouble of creating all these different typedefs for what are essentially SIZE_T. I'm fine with making them plain IntPtr/UIntPtr.

**Reply by kennykerr** [reply]
>  I would rather attribute individual fields to say if they're only in one arch or the other if we're not going to duplicate struct names.

The main benefit is that the nested struct can carry all of the differences in one place, not only fields, but also things like packing.

**Reply by kennykerr** [reply]
Parsers are also considerably harder to write when there are duplicate type names within a namespace.

**Reply by sotteson1** [reply]
> > I would rather attribute individual fields to say if they're only in one arch or the other if we're not going to duplicate struct names.
> 
> The main benefit is that the nested struct can carry all of the differences in one place, not only fields, but also things like packing.

Each arch-specific struct (like arch x64, arch x86) would need to contain the entire struct, right? Because you might have 3 common fields, then an x64-only field, then some more common fields, then an x86-only field, more common fields, etc. The CONTEXT struct does this kind of thing.

**Reply by kennykerr** [reply]
Yes, that's what I would expect.

**Reply by AArnott** [reply]
I like an approach that doesn't require special checks on each individual field, and doesn't cause us to have to check for the pack size from two distinct attributes. The winning ideas here IMO are an empty struct with the original name but attributes that point to the arch-specific variants (having garbage psuedo-names each), or the nested struct idea.

**Reply by kennykerr** [reply]
Yes, I don't really care if its nested or not.

**Comment by sotteson1** [comment]
How does this look? I like the ArchSupported attribute because there are sometimes methods/structs that only appear on certain architectures and then there aren't versions for the other architectures, but the ArchSpecificVersions leads you to the others ones if they exist. I could also combine ArchSupported and ArchSpecific, like:  ````[ArchSupported(Arch.x64 | Arch.arm64, x86 = "JSCreateContext__x86__")]````

````C#
    [ArchSupported(Arch.x64 | Arch.arm64)]
    [DllImport("USER32", ExactSpelling = true, SetLastError = true)]
    public static extern IntPtr GetWindowLongPtrA([In] HWND hWnd, [In] WINDOW_LONG_PTR_INDEX nIndex);

    [ArchSupported(Arch.x64 | Arch.arm64)]
    [ArchSpecific(x86 = "JSCreateContext__x86__")]
    [DllImport("chakra", ExactSpelling = true)]
    public unsafe static extern JsErrorCode JsCreateContext([In] void* runtime, [In] IDebugApplication64 debugApplication, [Out] void** newContext);

    [ArchSupported(Arch.x86)]
    [DllImport("chakra", EntryPoint = "JSCreateContext")]
    public unsafe static extern JsErrorCode JSCreateContext__x86__([In] void* runtime, [In] IDebugApplication32 debugApplication, [Out] void** newContext);

    [ArchSupported(Arch.x64)]
    [ArchSpecific(x86 = "CONTEXT__x86__", arm64 = "CONTEXT__arm64__")]
    public partial struct CONTEXT
    {
...
    }

    [ArchSupported(Arch.x86)]
    public partial struct CONTEXT__x86__
    {
...
    }

````

**Reply by AArnott** [reply]
Does this mean you would attribute *every* API? Or just the ones that are absent for at least *one* architecture?

I'm not a fan of the way you're using the attributes where one particular architecture gets preferred naming and the others are alternates. That creates an asymmetry that we'll have to undo in the projection and that complicates testing.
Instead, can *every* method have a unique name perhaps but an attribute that identifies its preferred name (or, just give them all the same preferred name since overloads are allowed in IL anyway). Then for the struct, perhaps just define the one empty struct with the preferred name and attributes to point projections to the alternates, each of which implement one of the arch-specific definitions.

**Reply by sotteson1** [reply]
For sure we don't want attributes on everything because it would bloat the winmd. If there are no attributes it implies it works for all architectures.

Metadata will let us duplicate method names and it will also let us duplicate struct names. It seems odd to not have methods and structs follow the same pattern, but we can do it that way if it's better for the projections. Maybe we don't want to duplicate the correctly-named struct because of typerefs from other types.

New proposals according to what you're suggesting.

Methods for different architectures share the same names:

````C#
    // Supported on all architectures
    [DllImport("USER32", ExactSpelling = true, SetLastError = true)]
    public static extern void Foo();

    // No x86 version supported
    [ArchSupported(Arch.x64 | Arch.arm64)]
    [DllImport("USER32", ExactSpelling = true, SetLastError = true)]
    public static extern IntPtr GetWindowLongPtrA([In] HWND hWnd, [In] WINDOW_LONG_PTR_INDEX nIndex);

    [ArchSupported(Arch.x64 | Arch.arm64)]
    [DllImport("chakra", ExactSpelling = true)]
    public unsafe static extern JsErrorCode JsCreateContext([In] void* runtime, [In] IDebugApplication64 debugApplication, [Out] void** newContext);

    [ArchSupported(Arch.x86)]
    [DllImport("chakra", EntryPoint = "JSCreateContext")]
    public unsafe static extern JsErrorCode JSCreateContext([In] void* runtime, [In] IDebugApplication32 debugApplication, [Out] void** newContext);

    [ArchSpecific(x64 = "CONTEXT__x64__", x86 = "CONTEXT__x86__", arm64 = "CONTEXT__arm64__")]
    public partial struct CONTEXT
    {
...
    }

    [ArchSupported(Arch.x64, Name = "CONTEXT")]
    public partial struct CONTEXT__x64__
    {
...
    }

    [ArchSupported(Arch.x86, Name = "CONTEXT")]
    public partial struct CONTEXT__x86__
    {
...
    }

````

or with arch-specific methods each having a unique name like the structs...

````C#
    // Supported on all architectures
    [DllImport("USER32", ExactSpelling = true, SetLastError = true)]
    public static extern void Foo();

    // No x86 version supported
    [ArchSupported(Arch.x64 | Arch.arm64)]
    [DllImport("USER32", ExactSpelling = true, SetLastError = true)]
    public static extern IntPtr GetWindowLongPtrA([In] HWND hWnd, [In] WINDOW_LONG_PTR_INDEX nIndex);

    [ArchSupported(Arch.x64 | Arch.arm64, Name = "JsCreateContext")]
    [DllImport("chakra", ExactSpelling = true)]
    public unsafe static extern JsErrorCode JsCreateContext__x64___arm64__([In] void* runtime, [In] IDebugApplication64 debugApplication, [Out] void** newContext);

    [ArchSupported(Arch.x86, Name = "JSCreateContext")]
    [DllImport("chakra", ExactSpelling = true)]
    public unsafe static extern JsErrorCode JSCreateContext__x86__([In] void* runtime, [In] IDebugApplication32 debugApplication, [Out] void** newContext);

    [ArchSpecific(x64 = "CONTEXT__x64__", x86 = "CONTEXT__x86__", arm64 = "CONTEXT__arm64__")]
    public partial struct CONTEXT
    {
...
    }

    [ArchSupported(Arch.x64, Name = "CONTEXT")]
    public partial struct CONTEXT__x64__
    {
...
    }

    [ArchSupported(Arch.x86, Name = "CONTEXT")]
    public partial struct CONTEXT__x86__
    {
...
    }

````

**Reply by mikebattista** [reply]
Can we rename ArchSupported to SupportedArchitecture to better align with SupportedOSPlatform?

**Reply by AArnott** [reply]
What about `amd64`? Do we need to distinguish between `x64` (aka IA64, Intel's less common arch) and `amd64`?

> It seems odd to not have methods and structs follow the same pattern, but we can do it that way if it's better for the projections.

That's a fair point. I am not sure how much my projection relies on name matching any more. I know it does at places, but I've been working to remove them. Maybe other projection authors should chime in here. If it wasn't expected to be problematic, perhaps retaining the original name *would* be simpler to explain and implement.

In your example:

```cs
    [ArchSpecific(x64 = "CONTEXT__x64__", x86 = "CONTEXT__x86__", arm64 = "CONTEXT__arm64__")]
    public partial struct CONTEXT
    {
...
    }
```

Your `...` is actually literally empty, correct? Is the idea that this struct would not represent anything, but stand in for the proper name and link to the arch-specific implementations?

> or with arch-specific methods each having a unique name like the structs...

👎 Nothing in metadata refers to the methods, so that just complicates things with no gain. The reason it *may* be useful for the structs is that typerefs to them exist, which a projection may still be trying to resolve by name. In fact, IIRC *all* type *refs* are in fact by name. So if we were to define multiple non-uniquely name structs, you had better be sure everywhere the metadata refers to them that it uses type**def** handles rather than type**ref** handles or projections really would have a problem.

**Reply by kennykerr** [reply]
TypeRefs are extremely common - the only way to resolve a TypeRef to its TypeDef is by name - so if there are duplicates it cannot be done unambiguously.

**Reply by kennykerr** [reply]
Oh the other hand, if we know there will only ever be one struct/function with that name for a given architecture we can filter out any others automatically. So assuming there are no overloads within architectures, we can avoid the attributes and mangled names. That seems simpler.

**Reply by AArnott** [reply]
> we can avoid the attributes and mangled names

How would we avoid *both*? If we have overloads and assume it's one-per arch, how do we know which arch an overload belongs to without either attributes or mangled names?

**Reply by kennykerr** [reply]
Sorry, I meant the _gratuitous_ attributes. So instead of:

```C#
    [ArchSpecificRef(x64 = "CONTEXT__x64__", x86 = "CONTEXT__x86__", arm64 = "CONTEXT__arm64__")]
    public struct CONTEXT
    {
        // empty
    }
    [ArchSpecificDef]
    public partial struct CONTEXT__x64__
    {
        // stuff
    }
    [ArchSpecificDef]
    public partial struct CONTEXT__x86__
    {
        // stuff
    }
    [ArchSpecificDef]
    public partial struct CONTEXT__arm64__
    {
        // stuff
    }
```

We could simply have:

```C#
    [Arch(x64)
    public struct CONTEXT
    {
        // stuff
    }
    [Arch(x86)
    public struct CONTEXT
    {
        // stuff
    }
    [Arch(arm64)
    public struct CONTEXT
    {
        // stuff
    }
```

**Reply by AArnott** [reply]
Right, although as I mention earlier, that means the metadata had better not use typeref's to reference the struct anywhere. It would have to use typedef handles instead. I don't know how feasible that is. 
Or maybe you're saying the projection would notice that multiple possible targets exist and then automatically disambiguate. I suppose they could, but that sounds like a slower and more complicated process. And it certainly violates how ECMA-335 was designed. Sure, a metadata writer/reader might not fail on it, but *clearly* if typeref's are only as precise as fully-qualified name, then surely it was never conceived to support name collisions like this.

**Reply by kennykerr** [reply]
Yes, this assumes you know how to filter out the overloads that don't match the target architecture. Assuming ECMA-335 can handle this, e.g. does ILASM let you write this, then it would seems _much_ simpler than having to use (1) meta structs with (2) complicated attributes with pointers to other (3) arch-specific structs.

**Reply by AArnott** [reply]
> this assumes you know how to filter out the overloads that don't match the target architecture

To be clear, have we switched to discussing methods? Methods may have overloads. Types do not. I thought we were talking about the structs.

ECMA-335 can handle method overloads easily, even where those overload signatures only vary by return type (which C# does not permit). A methodref evidently retains all signature information.

Contrast this with a typeref, which as you know only retains the full name (and scope). That's why I'm leaning against non-unique struct names.

**Reply by kennykerr** [reply]
Yes, method overloads are supported - WinRT uses them liberally. 

Yes, struct overloads seem unusual/unsupported, but Steve seemed to imply that they are possible. And if you can guarantee that there are no ambiguities then the `TypeRef` to `TypeDef` mapping would work. All I'm pointing out is that if they are possible then it might be a lot simpler that way. 

I'm not sold on any one solution, but some of the alternatives above just seem very complicated.

**Reply by sotteson1** [reply]
I'm pretty sure I can have duplicate struct names, because that's a bug I've had before that my tests now catch. The problem is that the typerefs that use an arch-specific struct can only point at one of them. GetOpenFileNameW (same signature for all architectures) takes OPENFILENAMEW which has different packing for 64 and 32 bit. So a projection would see the typeref to OPENFILENAMEW and then have to notice that it's arch-specific and then look for the other type with the same name, but with the desired architecture.

**Reply by kennykerr** [reply]
Ignoring scope, a `TypeRef` only stores a namespace/name pair so it doesn't point to any specific definition. It merely provides a namespace/name that can be used to look up a `TypeDef`. 

![image](https://user-images.githubusercontent.com/9845234/115474209-d2289f80-a1f1-11eb-99c1-aa1ab4a56112.png)

**Reply by sotteson1** [reply]
Ah, cool! So a projection would have to know there could be more than one struct that matches namespace+name, which is still kind of weird, and then look to match any arch-specific versions. Methods would work the same way.

**Reply by AArnott** [reply]
> Methods would work the same way.

Well, nothing references a method so there's no matching required for that. Except I guess your attributes like `RIAA`.

I'm still not a fan of deliberately using non-unique type names for reasons I gave above, but I think I can deal with it.
